daemon/__init__.py

```diff
@@ -160,17 +151,43 @@ def _start_consumer():
     ConsumerThread().start()
 
 
-def _update_default_args():
-    global jinad_args, __root_workspace__
-    jinad_args = _get_run_args()
-    __root_workspace__ = '/workspace' if jinad_args.mode else jinad_args.workspace
+def _start_uvicorn(app: 'FastAPI'):
+    config = Config(
+        app=app,
+        host=jinad_args.host,
+        port=jinad_args.port_expose,
+        loop='uvloop',
+        log_level='error',
+    )
+    server = Server(config=config)
+    server.run()
 
 
-def main():
-    """Entrypoint for jinad"""
+def setup():
+    """Setup steps for JinaD"""
     _update_default_args()
     pathlib.Path(__root_workspace__).mkdir(parents=True, exist_ok=True)
     if not jinad_args.no_fluentd:
         Thread(target=_start_fluentd, daemon=True).start()
     _start_consumer()
     _start_uvicorn(app=_get_app(mode=jinad_args.mode))
+
+
+def teardown():
```

Comment:
Shouldn't we stop the consumer thread and uvicorn here?

Reply:
Consumer thread is a daemon thread, so no need to close it manually.
Uvicorn handles its own `shutdown` logic in `serve` itself for different signal handlers.